execinfra: ensure all processors are properly cleaned up on shutdown

yuzefovich opened the pull request:
This commit extends `Processor` interface to introduce `Close` method which must be called at least once. For processors that implement `RowSource` interface this has already been the case with `ConsumerClosed` method, but a handful of processors (sampler, sample aggregator, and index backfiller) didn't have a way to release its resources in case the flow is set up but never ran (which can happen on the server shutdown). This issue is now fixed.

Fixes: #126744.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127020)
<!-- Reviewable:end -->

yuzefovich committed:
execinfra: ensure all processors are properly cleaned up on shutdown

This commit extends `Processor` interface to introduce `Close` method
which must be called at least once. For processors that implement
`RowSource` interface this has already been the case with
`ConsumerClosed` method, but a handful of processors (sampler, sample
aggregator, and index backfiller) didn't have a way to release its
resources in case the flow is set up but never ran (which can happen on
the server shutdown). This issue is now fixed.

Release note: None

DrewKimball reviewed (approved):
Nice find! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127020#-:-O1ZFvhT7dV28z5LqZ3Q:bli3xsu)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
